File: src/behavior_selection.m
function validBhv = behavior_selection(data, opts)
% Get indices of usable behaviors
% Details:
%    None
%
% Syntax:
%     validBhv = behavior_selection(data, opts)
%
% Inputs:
%    data - 
%    opts - 
%
% Outputs:
%    validBhv - [m,n] size,[double] type,Description
%
% Example:
%    None
%
% See also: None


codes = unique(data.ID);
behaviors = {};
for iBhv = 1 : length(codes)
    firstIdx = find(data.ID == codes(iBhv), 1);
    behaviors = [behaviors, data.Name{firstIdx}];
    % fprintf('behavior %d:\t code:%d\t name: %s\n', i, codes(i), dataBhvAlex.Behavior{firstIdx})
end
validBhv = zeros(size(data, 1), 1);

for i = 1 : length(codes) % length(actList)

    iAct = codes(i);

    actIdx = data.ID == iAct; % All instances labeled as this behavior
    allPossible = sum(actIdx);

    longEnough = data.Dur >= opts.minActTime; % Only use if it lasted long enough to count

    actAndLong = actIdx & longEnough;
    andLongEnough = sum(actAndLong);  % for printing sanity check report below

    % iPossible is a list of behavior indices for this behavior that is
    % at least long enough
    % Go through possible instances and discard unusable (repeated) ones
    for iPossible = find(actAndLong)'

        % Was there the same behvaior within the last minNoRepeat sec?
        endTime = [data.StartTime(2:end); data.StartTime(end) + data.Dur(end)];
        % possible repeated behaviors are any behaviors that came
        % before this one that were within the no-repeat minimal time
        iPossRepeat = endTime < data.StartTime(iPossible) & endTime >= (data.StartTime(iPossible) - opts.minNoRepeatTime);

        % sanity checks
        % preBehv = sum(iPossRepeat);


        % If it's within minNoRepeat and any of the behaviors during that time are the same as this one (this behavior is a repeat), get rid of it
        if sum(iPossRepeat) && any(data.ID(iPossRepeat) == iAct)

            % % debug display
            % data.bStart100(iPossible-3:iPossible+3,:)
            % removeTrial = iPossible

            actAndLong(iPossible) = 0;

        end
    end



    andNotRepeated = sum(actAndLong);

    fprintf('%d: %s: Valid: %d\t (%.1f)%%\n', codes(i), behaviors{i}, andNotRepeated, 100 * andNotRepeated / allPossible)

    validBhv(actAndLong) = 1;
end

File: src/behavior_selection.py
import numpy as np

def behavior_selection(data, opts):
    behaviors = opts.behaviors
    codes = opts.bhvCodes

    # validBhv = [None] * len(behaviors)
    validBhv = np.zeros((data.shape[0], len(behaviors)))

    for i in range(len(codes)):
        if codes[i] in opts.validCodes:
            iAct = codes[i]
            actIdx = data.bhvID == iAct  # All instances labeled as this behavior
            allPossible = sum(actIdx)

            longEnough = data.bhvDur >= opts.minActTime  # Only use if it lasted long enough to count

            actAndLong = actIdx & longEnough
            andLongEnough = sum(actAndLong)  # for printing sanity check report below

            # iPossible is a list of behavior indices for this behavior that is
            # at least long enough
            # Go through possible instances and discard unusable (repeated) ones
            for iPossible in np.where(actAndLong)[0]:
                # Was there the same behavior within the last minNoRepeat sec?
                # endTime = np.concatenate([data.bhvStartTime[1:], [data.bhvStartTime[-1] + data.bhvDur[-1]]])
                endTime = np.concatenate([data['bhvStartTime'].iloc[1:], [data['bhvStartTime'].iloc[-1] + data['bhvDur'].iloc[-1]]])
                # possible repeated behaviors are any behaviors that came
                # before this one that were within the no-repeat minimal time
                iPossRepeat = (endTime < data.bhvStartTime[iPossible]) & (endTime >= (data.bhvStartTime[iPossible] - opts.minNoRepeatTime))

                # If it's within minNoRepeat and any of the behaviors during that time are the same as this one (this behavior is a repeat), get rid of it
                if np.sum(iPossRepeat) and any(data.bhvID[iPossRepeat] == iAct):
                    actAndLong[iPossible] = False

            andNotRepeated = sum(actAndLong)

            print(f'Behavior {codes[i]}: {behaviors[i]}')
            print(f'{allPossible}: allPossible')
            print(f'{andLongEnough}: andLongEnough')
            print(f'{andNotRepeated}: andNotRepeated')
            print(f'Percent valid: {100 * andNotRepeated / allPossible:.1f}\n')

            if sum(actAndLong) >= opts.minBhvNum:
                validBhv[:,i] = actAndLong
            else:
                validBhv[:,i] = np.full(len(actAndLong), False)
                print(f'Not enough {behaviors[i]} bouts to analyze ({sum(actAndLong)} of {opts.minBhvNum} needed)\n')

        else:
            validBhv[:,i] = np.full(len(data), False)
            print(f'{behaviors[i]} code {codes[i]} is not a valid behavior for this analysis\n\n')

    return validBhv
